Yoast/javascript
# Prevent abbreviations from stemming

## Summary

* Adds functionality that processes abbreviations (2-4 letter long upper-case words) differently than the rest of the words in internal linking suggestions and insights. Abbreviations do not get stemmed, i.e., they are saved in the database as is. Abbreviations are displayed in the Insights as upper-case words.

## Relevant technical choices:

* For internal linking abbreviations are being determined based on all word occurrences in the text and in the attributes (title, keyphrase, synonyms, metadescription). For insights abbreviations are determined based on the text alone.

## Test instructions

This PR can be tested by following these steps:

* Use the current branch and the same-name branch on `premium-configuration`.
* Check if all tests pass.
* Use `example` to test that
  * A text without abbreviations gets processed as expected
  * In a text with abbreviations:
    * All non-abbreviations are processed in the old way (get stemmed for the `stem` field, get lowercased for the `word` field)
    * All abbreviations are processed differently (don't get stemmed, get uppercased in the `word` field)
  * Try editing the text though adding and removing words, make sure the analysis does not crashes and the results in the `Prominent words` fields make sense.
* Build a local version of the premium plugin (if [this PR](https://github.com/Yoast/wordpress-seo-premium/pull/2296) has been merged)
  * Run the dev server to see debug messages and print-out of results of researches
  * Use branch `feature/internal-linking` on `wordpress-seo-premium`.
  * **Not-so-nifty-trick**: Because of the fact that we cannot use custom versions of the `premium-configuration` when building the plugin locally, you will need to disable all functionality that relies on the rules added in the `feature/internal-linking`.
    * In `/yoastseo/src/morphology/english/determineStem.js` comment out lines 96 to 99
    * In `/yoastseo/src/morphology/english/getAdjectiveForms.js` comment out lines 82 to 112
  * Link `yoastseo` and `yoast-components` to your `wordpress-seo-premium` branch and build `javascript`, `wordpress-seo-premium` and the `premium` folder of the latter.
  * Open the local version of the plugin and add a new post.
  * Note down the `id` of the post and check which words get saved in the database. To see the words that get saved, you can use a query 
```
SELECT * FROM `wp_term_relationships`
INNER JOIN `wp_term_taxonomy` ON `wp_term_taxonomy`.`term_taxonomy_id` = `wp_term_relationships`.`term_taxonomy_id`
INNER JOIN `wp_terms` ON `wp_term_taxonomy`.`term_id` = `wp_terms`.`term_id`
WHERE `object_id` = 'XXXXXXXXXXX';
```
*XXXXXXXXXXX should be your post id*
  * Check that
    * A text without abbreviations gets processed as expected
    * In a text with abbreviations:
      * All non-abbreviations are processed in the old way (get stemmed for the `stem` field, get lowercased for the `word` field)
      * All abbreviations are processed differently (don't get stemmed, get uppercased in the `word` field)

## UI changes

## Quality assurance


Fixes #139
